### PR TITLE
Issue 728 gr oresidwrap

### LINKED ISSUE #728: Large .gro files (>100000 residues) have incorrect residue numbering
I have .gro file with over 99999 residues, so there's a point where it looks like this:

```
99993W        W24409   5.197  38.171   2.358  0.0392  0.2897  0.1634
99994W        W24410  46.548  35.562   3.925  0.0041 -0.0044 -0.2686
99995W        W24411  41.678  11.567   4.117 -0.0489  0.2279  0.0398
99996W        W24412  32.466  10.908   2.439  0.0176  0.0328  0.2482
99997W        W24413  12.719  46.133   1.389  0.3063 -0.2728 -0.1758
99998W        W24414  15.130  10.483   9.969 -0.0133  0.2911  0.0727
99999W        W24415  11.716  36.838   1.308 -0.0390  0.1237  0.0310
    0W        W24416  40.024  47.039  11.701  0.0180 -0.0193  0.1456
    1W        W24417   8.068  42.350  10.090  0.1637 -0.0463  0.0713
    2W        W24418   4.284  28.808  10.156  0.1226  0.4642  0.0802
    3W        W24419  23.977  40.007  10.608 -0.0688  0.1545  0.1058
    4W        W24420  43.318  30.071   9.808 -0.1199 -0.1253 -0.1616
    5W        W24421   6.697  24.131  11.544 -0.0203  0.2435 -0.2320
    6W        W24422   1.650  36.335  10.132  0.0300 -0.1128 -0.3979
    7W        W24423   0.771  17.546   5.260  0.1603  0.0087  0.1779
    8W        W24424  49.619   4.969   9.924 -0.0225  0.0800 -0.4119
    9W        W24425  51.452  43.795   9.936  0.2769  0.0756 -0.1641
```

If I load this into MDAnalysis, the atom numbering is correct, however residue numbering is incorrect - as you can see below, residue 1 now includes a water molecule, which is incorrectly numbered as 1 (should be 100001).

```
    >>> universe.residues[0]
    <Residue POPE, 1>
    >>> universe.residues[0].atoms[-1]
    <Atom 324417: W of type W of resname W, resid 1 and segid SYSTEM>
    >>> u.atoms[224416].residue
    <Residue POPE, 1>
    >>> u.atoms[224416].resid
    1
```

Can I propose the following correction to `mdanalysis/package/MDAnalysis/topology/GROParser.py
` (line 45 onwards, new code surrounded by ###triple hashes###):

``` python
class GROParser(TopologyReader):
    format = 'GRO'

    def parse(self):
        """Parse GRO file *filename* and return the dict `structure`.
        Only reads the list of atoms.
        :Returns: MDAnalysis internal *structure* dict
        .. SeeAlso:: The *structure* dict is defined in
                     :func:`MDAnalysis.topology.base`.
        """
        segid = "SYSTEM"
        atoms = []
        with openany(self.filename, "rt") as grofile:
            grofile.readline()
            natoms = int(grofile.readline())
            ### ADDED CODE
            resid_correction = 0
            ### 
            for atom_iter in range(natoms):
                line = grofile.readline()
                try:
                    resid, resname, name = int(line[0:5]), line[5:10].strip(), line[10:15].strip()
                    ### ADDED CODE
                    resid = int(resid)
                    if resid == 0:
                        resid_correction += 100000
                    resid += resid_correction
                    ###
                    # guess based on atom name
                    elem = guess_atom_element(name)
                    atype = elem
                    mass = get_atom_mass(elem)
                    charge = guess_atom_charge(name)
                    # segid = "SYSTEM"
                    # ignore coords and velocities, they can be read by coordinates.GRO
                except:
                    raise IOError("Couldn't read the following line of the .gro file:\n"
                                  "{0}".format(line))
                else:
                    # Just use the atom_iter (counting from 0) rather than
                    # the number in the .gro file (which wraps at 99999)
                    atoms.append(Atom(atom_iter, name, atype, resname, resid,
                                      segid, mass, charge, universe=self._u))
        structure = {'atoms': atoms}

        return structure
```

(sorry for the lack of a proper pull request - I'm pushed for time today but I'll figure out how to do it in future)

### PR DESCRIPTION
**Pull request by richardjgowers**
Fixes #728

PR Checklist
------------
 - [x] Tests?
 - [ ] Docs?
 - [x] CHANGELOG updated?
 - [x] Issue raised/referenced?

### CODE REVIEW

**Comment by richardjgowers** [comment]
@jbarnoud the fix in #728 only actually worked for size 1 residues, otherwise **every** time it saw resid=0 it incremented the resid_correction

File: package/MDAnalysis/topology/GROParser.py [hunk 1]
```diff
@@ -89,8 +89,9 @@ def parse(self):
             names = np.zeros(n_atoms, dtype=object)
             indices = np.zeros(n_atoms, dtype=np.int32)
 
-            for i in range(n_atoms):
-                line = inf.readline()
+            for i, line in enumerate(inf):
```
Comment (jbarnoud) [inline package/MDAnalysis/topology/GROParser.py]: Thanks! This looks much nicer than before.

**Comment by richardjgowers** [comment]
thanks @jbarnoud , not sure why it would do that, but I can see it's a corner case I didn't think of :)  So I guess we want to ignore any 0-resid blocks if they're the first set of resids?

**Comment by jbarnoud** [comment]
If the first residue is 0, then it should probably stay 0.

**Comment by richardjgowers** [comment]
@jbarnoud fixed your corner case, I challenge you to break it now! :P

**Pull request merged**